enhancements/update/cvo-log-level-api.md

```diff
@@ -0,0 +1,194 @@
+
+```go
+// +kubebuilder:validation:Enum="";Normal;Debug;Trace;TraceAll
+type CVOLogLevel string
+
+var (
+	// Normal is the default.  Normal, working log information, everything is fine, but helpful notices for auditing or common operations.  In kube, this is probably glog=2.
+	CVOLogLevelNormal CVOLogLevel = "Normal"
+
+	// Debug is used when something went wrong.  Even common operations may be logged, and less helpful but more quantity of notices.  In kube, this is probably glog=4.
+	CVOLogLevelDebug CVOLogLevel = "Debug"
+
+	// Trace is used when something went really badly and even more verbose logs are needed.  Logging every function call as part of a common operation, to tracing execution of a query.  In kube, this is probably glog=6.
+	CVOLogLevelTrace CVOLogLevel = "Trace"
+
+	// TraceAll is used when something is broken at the level of API content/decoding.  It will dump complete body content.  If you turn this on in a production cluster
+	// prepare from serious performance issues and massive amounts of logs.  In kube, this is probably glog=8.
+	CVOLogLevelTraceAll CVOLogLevel = "TraceAll"
+)
+```
+
+### Workflow Description
+
+Given a cluster administrator and a working cluster for which the administrator is responsible.
+
+**cluster administrator** is a human user responsible for managing the cluster.
+
+1. The cluster administrator notices an issue in the cluster and chooses to troubleshoot the issue.
+2. The cluster administrator after some troubleshooting notices that the logs of the Cluster Version Operator (CVO) might help.
+3. The cluster administrator notices that the logs are not detailed enough to troubleshoot the issue, and the administrator raises the log level from the default value `Normal` to the level `Trace` by simply modifying the `ClusterVersion` resource via the web console or by patching the resource by using the CLI.
+4. The cluster administrator fixes the issue.
+5. The cluster administrator notices that the CVO outputs too many logs for the administrator's liking.
+6. The cluster administrator lowers the log level of the CVO to the level `Normal`.
+7. The cluster administrator is now a happy cluster administrator.
+
+#### Variation [optional]
+
+N/A
+
+### API Extensions
+
+See the section [Proposal](#proposal) for the proposed API changes.
+
+### Implementation Details/Notes/Constraints [optional]
+
+The CVO needs to change the log level dynamically without redeploying itself. The CVO currently uses the package [`klog`](https://github.com/kubernetes/klog) for logging. This package does not provide a simple way of changing the log level dynamically. However, it is possible, and there already exists a function 
+[`SetLogLevel`](https://github.com/openshift/library-go/blame/8477ec72b72554c01afff4df9c6a90c0e492ea87/pkg/operator/loglevel/util.go#L63-L103) in the [`openshift/library-go`](https://github.com/openshift/library-go) repository to remedy this issue.
+
+#### Hypershift [optional]
+
+N/A
```

Comment:
Possibly worth pointing out that there's currently no log-level knob in [the HostedControlPlane `spec`][1], and [they hard-code `--v=4`][2], so no changes proposed there at the moment.  But that if the HyperShift folks want access to this kind of functionality, they'd presumably add a knob to the HostedControlPlane `spec`, likely fed from a new knob in the HostedCluster `spec`, and then propagate that value down to their CVO Deployment.  Or maybe they'd leave their Deployment default hard-coded to avoid Deployment rollouts and only drive the CVO's log level via ClusterVersion?

[1]: https://github.com/openshift/hypershift/blob/90aa44d064f6fe476ba4a3f25973768cbdf05eb5/api/v1beta1/hosted_controlplane.go#L26-L162
[2]: https://github.com/openshift/hypershift/blob/90aa44d064f6fe476ba4a3f25973768cbdf05eb5/control-plane-operator/controllers/hostedcontrolplane/cvo/reconcile.go#L335

Reply:
Hmm, actually, HyperShift might need to grow an opinion (and get a vendor bump to add the property) in [their `Update`-based ClusterVersion reconciler][1].  Unchanged, the `Update` logic would stomp off the new property, as the Go deserialization silently dropped the unknown property.  And after a vendor bump to pick up the new property, they'd pass along any pre-existing value, creating a configuration knob for management cluster behavior (CVO log level) that could be driven from the hosted-cluster admin (ClusterVersion `spec`).  They probably want to at least ensure the new property is getting stomped to an empty string, until they optionally grow HostedControlPlane/HostedCluster `spec` knobs around it in the future.

[1]: https://github.com/openshift/hypershift/blob/90aa44d064f6fe476ba4a3f25973768cbdf05eb5/control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go#L972-L978

Reply:
I have missed this, thank you! I have updated the enhancement, and I have proposed to reconcile a fixed value for this moment. When the PR is marked as _Ready for review_, HyperShift folks may give more feedback.